### ISSUE TITLE
Rule proposal: Boolean variable naming

### ISSUE THREAD

**Issue by sindresorhus** [typescript-eslint/typescript-eslint#515]
I would like to enforce that variables and properties with a boolean use the prefix `is`, `has`, `can`, `did`, and `will`. For example: `.visible` → `.isVisible`.

I often see `.visible`, `.open`, etc, in code, and it's not clear whether they are booleans, contain something visible/open, or whether they contain a function that makes something visible or opens something.

## Fail

```ts
const visible = true;

const rainbow = false;
```

## Pass

```ts
const isVisible = true;

const hasRainbow = false;
```

**Comment by sw-yx** [comment]
i often use past/future tense stuff for dirty-flag type operations - thoughts? `didChange`, `willCleanup`.

**Comment by sindresorhus** [comment]
@sw-yx Ah yes, forgot about those. They make sense too.

**Comment by platinumazure** [comment]
Seems this could be generalized as sort of like core `id-match`, but applied to specific types. Configuration could look something like:

```json
{
    "typescript/typed-id-match": [
        "error",
        { "types": [/* ... */], "pattern": "^(is|has|can|did|will)$" }
    ]
}
```

**Comment by golopot** [comment]
Should this rule deal with properties? How to deal with property names you cannot control, for example, other people's api?
```js
// it probably does not make sense to warn on this
const options = {capture: true, passive: false};
someExternalApi(options);
```

**Comment by golopot** [comment]
Does this rule cover boolean valued functions?

**Comment by bradzacher** [comment]
There's no reason the rule couldn't support all of these things and have options to enable/disable checks.

**Comment by loilo** [comment]
Side note: Another possible prefix that may be considered is `should`. This is very semantically similar to `will`, but in some cases I find `should` the more fitting choice (e.g. when used as a function parameter).

As an example from the code that induced this comment:

```ts
function formatArray(data: any[], shouldIgnoreEmpty = false) {
  if (data.length === 0 && shouldIgnoreEmpty) {
    return ''
  } else {
    return JSON.stringify(data)
  }
}
```
